src/metadata/ffmpeg_thumbnailer_handler.cc

```diff
@@ -0,0 +1,142 @@
+    along with Gerbera.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/// \file ffmpeg_thumbnailer_handler.cc
+
+#ifdef HAVE_FFMPEGTHUMBNAILER
+
+#include "ffmpeg_thumbnailer_handler.h"
+#include <libffmpegthumbnailer/filmstripfilter.h>
+#include <libffmpegthumbnailer/videothumbnailer.h>
+
+fs::path FfmpegThumbnailerHandler::getThumbnailCacheBasePath(const Config& config_)
+{
+    auto configuredDir = config_.getOption(CFG_SERVER_EXTOPTS_FFMPEGTHUMBNAILER_CACHE_DIR);
+    if (!configuredDir.empty())
+        return configuredDir;
+
+    auto home = config_.getOption(CFG_SERVER_HOME);
+    return fs::path(home) / "cache-dir";
+}
+
+fs::path FfmpegThumbnailerHandler::getThumbnailCachePath(const fs::path& base, const fs::path& movie)
+{
+    assert(movie.is_absolute());
+
+    auto path = base / movie.relative_path();
+    path += "-thumb.jpg";
+    return path;
+}
+
+std::optional<std::vector<std::byte>> FfmpegThumbnailerHandler::readThumbnailCacheFile(const fs::path& movieFilename) const
+{
+    auto path = getThumbnailCachePath(getThumbnailCacheBasePath(*config), movieFilename);
+    return GrbFile(path).readBinaryFile();
+}
+
+void FfmpegThumbnailerHandler::writeThumbnailCacheFile(const fs::path& movieFilename, const std::byte* data, std::size_t size) const
+{
+    try {
+        auto path = getThumbnailCachePath(getThumbnailCacheBasePath(*config), movieFilename);
+        fs::create_directories(path.parent_path());
+        GrbFile(path).writeBinaryFile(data, size);
+    } catch (const std::runtime_error& e) {
+        log_error("Failed to write thumbnail cache: {}", e.what());
+    }
+}
+
+std::unique_ptr<IOHandler> FfmpegThumbnailerHandler::serveContent(const std::shared_ptr<CdsObject>& obj, int resNum)
+{
+    auto item = std::dynamic_pointer_cast<CdsItem>(obj);
+    if (!item)
+        return nullptr;
+
+    if (!config->getBoolOption(CFG_SERVER_EXTOPTS_FFMPEGTHUMBNAILER_ENABLED))
+        return nullptr;
+
+    auto cacheEnabled = config->getBoolOption(CFG_SERVER_EXTOPTS_FFMPEGTHUMBNAILER_CACHE_DIR_ENABLED);
+    if (cacheEnabled) {
+        auto data = readThumbnailCacheFile(item->getLocation());
+        if (data) {
+            log_debug("Returning cached thumbnail for file: {}", item->getLocation().c_str());
+            return std::make_unique<MemIOHandler>(data->data(), data->size());
+        }
+    }
+
+    auto thumbLock = std::scoped_lock<std::mutex>(thumb_mutex);
+
+    auto th = ffmpegthumbnailer::VideoThumbnailer(config->getIntOption(CFG_SERVER_EXTOPTS_FFMPEGTHUMBNAILER_THUMBSIZE), false, true, config->getIntOption(CFG_SERVER_EXTOPTS_FFMPEGTHUMBNAILER_IMAGE_QUALITY), false);
+    std::vector<uint8_t> img;
+
+    th.setSeekPercentage(config->getIntOption(CFG_SERVER_EXTOPTS_FFMPEGTHUMBNAILER_SEEK_PERCENTAGE));
+    if (config->getBoolOption(CFG_SERVER_EXTOPTS_FFMPEGTHUMBNAILER_FILMSTRIP_OVERLAY))
+        th.addFilter(new ffmpegthumbnailer::FilmStripFilter());
```

Comment:
I can't tell if this causes a memory leak. Where's the corresponding delete?

Reply:
This is the existing code, if you want to double check that it would be great.